backends/s3/s3.go

```diff
@@ -109,36 +109,29 @@ func (b *Backend) List(ctx context.Context, prefix string) (simpleblob.BlobList,
 	}
 
 	// Request and cache full list, and use marker file to invalidate the cache
-	now := time.Now()
 	data, err := b.Load(ctx, UpdateMarkerFilename)
-	if err != nil && !os.IsNotExist(err) {
-		return nil, err
+	exists := true
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		exists = false
 	}
 	current := string(data)
 
-	b.mu.Lock()
-	lastMarker := b.lastMarker
-	age := now.Sub(b.lastTime)
-	b.mu.Unlock()
-
-	var blobs simpleblob.BlobList
-	if current != lastMarker || age >= b.opt.UpdateMarkerForceListInterval {
+	if !exists || b.lastMarker == "" || current != b.lastMarker || time.Since(b.lastTime) >= b.opt.UpdateMarkerForceListInterval {
 		// Update cache
-		blobs, err = b.doList(ctx, "") // all, no prefix
+		blobs, err := b.doList(ctx, "") // all, no prefix
 		if err != nil {
 			return nil, err
 		}
-
 		b.mu.Lock()
 		b.lastMarker = current
 		b.lastList = blobs
-		b.mu.Unlock()
-	} else {
-		b.mu.Lock()
-		blobs = b.lastList
+		b.lastTime = time.Now()
 		b.mu.Unlock()
 	}
-	return blobs.WithPrefix(prefix), nil
+	return b.lastList.WithPrefix(prefix), nil
```

Comment:
This is unsafe. You need to return `blobs`, because `b.lastList` access requires holding the lock.